mirgecom/profiling.py

```diff
@@ -0,0 +1,229 @@
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from meshmode.array_context import PyOpenCLArrayContext
+import pyopencl as cl
+from pytools.py_codegen import PythonFunctionGenerator
+import loopy as lp
+import numpy as np
+from dataclasses import dataclass
+import pytools
+
+__doc__ = """
+.. autoclass:: PyOpenCLProfilingArrayContext
+"""
+
+
+@dataclass
+class ProfileResult:
+    """Class to hold the results of a single kernel execution."""
+
+    time: int
+    flops: int
+    bytes_accessed: int
+    footprint_bytes: int
+
+
+@dataclass
+class ProfileEvent:
+    """Class to hold a profile event that has not been seen by the profiler yet."""
+
+    cl_event: cl._cl.Event
+    program: lp.kernel.LoopKernel
+    kwargs: dict
+
+
+class PyOpenCLProfilingArrayContext(PyOpenCLArrayContext):
+    """An array context that profiles kernel executions.
+
+    .. automethod:: tabulate_profiling_data
+    .. automethod:: call_loopy
+
+    Inherits from :class:`meshmode.array_context.PyOpenCLArrayContext`.
+    """
+
+    def __init__(self, queue, allocator=None) -> None:
+        super().__init__(queue, allocator)
+
+        if not queue.properties & cl.command_queue_properties.PROFILING_ENABLE:
+            raise RuntimeError("Profiling was not enabled in the command queue. "
+                 "Please create the queue with "
+                 "cl.command_queue_properties.PROFILING_ENABLE.")
+
+        self.profile_events = []
+        self.profile_results = {}
+        self.kernel_stats = {}
+
+    def _finish_profile_events(self) -> None:
+        # First, wait for completion of all events
+        if self.profile_events:
+            cl.wait_for_events([pevt.cl_event for pevt in self.profile_events])
+
+        # Then, collect all events and store them
+        for t in self.profile_events:
+            program = t.program
+            r = self._get_kernel_stats(program, t.kwargs)
+            time = t.cl_event.profile.end - t.cl_event.profile.start
+
+            new = ProfileResult(time, r.flops, r.bytes_accessed, r.footprint_bytes)
+
+            self.profile_results.setdefault(program, []).append(new)
+
+        self.profile_events = []
+
+    def tabulate_profiling_data(self) -> pytools.Table:
+        """Return a :class:`pytools.Table` with the profiling results."""
+        self._finish_profile_events()
+
+        tbl = pytools.Table()
+
+        tbl.add_row(["Function", "Calls",
+            "Time_min [s]", "Time_avg [s]", "Time_max [s]",
+            "GFlops/s_min", "GFlops/s_avg", "GFlops/s_max",
+            "BWAcc_min [GByte/s]", "BWAcc_mean [GByte/s]", "BWAcc_max [GByte/s]",
+            "BWFoot_min [GByte/s]", "BWFoot_mean [GByte/s]", "BWFoot_max [GByte/s]"])
```

Comment:
I know this is already a pretty column-heavy table, but I *think* it could be useful to also print estimates of intensity (i.e. flops/byte).

Reply:
Added (just the mean for now).